### PR TITLE
Rethink penalty manager

### PR DESCRIPTION
**Pull request by leonlan**
I made a PR so I can review the changes on this branch already.

### CODE REVIEW

File: pyvrp/PenaltyManager.py
Comment (leonlan) [inline pyvrp/PenaltyManager.py:1]: https://pyvrp--196.org.readthedocs.build/en/196/api/pyvrp.html

This doesn't show up in the docs

**Comment by leonlan** [comment]
Benchmark on 10 seeds GH1000 for this PR:
```
min     33358.6
mean    33375.9
max     33391.9
```

**Comment by wouterkool** [comment]
Thanks! It looks somewhat worse but I don't know if it's significant? Do you have the number of iterations as well?

**Comment by leonlan** [comment]
> Thanks! It looks somewhat worse but I don't know if it's significant? Do you have the number of iterations as well?

I'll run one for main now

**Comment by leonlan** [comment]
```
min     33352.2
mean    33371.8
max     33384.1
std         9.2
```
This is the result for main, so there's almost no difference between this PR and main. Some of the earlier merged PRs probably had some small impact on the performance, but that's not for this PR.

File: pyvrp/_CostEvaluator.pyi
Comment (N-Wouda) [inline pyvrp/_CostEvaluator.pyi:1]: Can we add this class to the docs? Something like
```sphinx
.. automodule:: pyvrp._CostEvaluator
   
   .. autoapiclass:: CostEvaluator
      :members:
```
in `docs/source/api/pyvrp.rst`.

File: pyvrp/tests/test_CostEvaluator.py
Comment (N-Wouda) [reply pyvrp/tests/test_CostEvaluator.py:1]: Oh those are in #199

File: pyvrp/cpp/Individual.cpp
Comment (N-Wouda) [inline pyvrp/cpp/Individual.cpp:185]:
```suggestion
    out << "Distance: " << indiv.distance() << '\n';
```

File: pyvrp/cpp/Individual.h
Comment (N-Wouda) [inline pyvrp/cpp/Individual.h:14]: This no longer has to be a friend now that `distance()` is available. Can we revert this to its previous position below the class?

File: pyvrp/cpp/Individual_bindings.cpp
Comment (N-Wouda) [inline pyvrp/cpp/Individual_bindings.cpp:26]:
```suggestion
                    py::arg("rng"));
```

I'm reasonably sure this was wrong before.

File: pyvrp/cpp/Individual.h
Comment (N-Wouda) [reply pyvrp/cpp/Individual.h:14]: (I'm pretty sure we can now "unfriend" `std::hash` as well, but that's out of scope for this PR)

**Review by N-Wouda — APPROVED** [review]
LGTM once #199 has been merged into this!

**Comment by wouterkool** [comment]
Closed this pull request in favor of #200 to get authorship right (to support git blame pointing to me for this). Also 200 is a nicer number.

**Pull request closed without merging**
